### PR TITLE
Metrics implementation

### PR DESCRIPTION
**Pull request by devrimcavusoglu**

### CODE REVIEW

File: trapper/metrics/metric.py
Comment (cemilcengiz) [inline trapper/metrics/metric.py:1]: I think it would be nice if there was a way to see the available metrics and their arguments without looking at jury's source code although I guess it is hard to do. Actually, with the datasets integration, we have a similar problem (kind of) but they have more than 100 datasets so it is OK for that. In the optimizers, model_wrappers etc, we have our wrappers sitting as a layer between trapper and allennlp (optimizers) and transformers (model_wrappers). What I would like to have a similar layer for the metrics. The problem is that Jury already has a nice class hierarchy with concrete subclasses as well as factories, therefore I do not like repeating the whole structure here in trapper. Nevertheless, I wanted to share this thought.

File: examples/question_answering/question_answering.ipynb
Comment (cemilcengiz) [inline examples/question_answering/question_answering.ipynb:9]: Can we add a requirements file for the question answering example as well?

File: README.md
Comment (cemilcengiz) [inline README.md:73]: Title words should be uppercase (except for "for" end "through" I guess)

File: README.md
Comment (cemilcengiz) [inline README.md:75]: Why don't we change the first sentence to something like:
"trapper supports the common NLP metrics through the jury library.".

File: README.md
Comment (cemilcengiz) [inline README.md:76]: The last sentence is more suitable in Jury source code IMHO. At least, lets remove the emoji :D

File: README.md
Comment (cemilcengiz) [inline README.md:100]: Either use `the` in jury library, or drop the `library` word. Also, I think we could write something likt `In order to make the metrics flexible enough to work with the trainer in a common interface` instead of `In order to fully utilize transformers' Trainer's metric computation`. Also, there should be a comma after this one regardless (before we introduced )

File: README.md
Comment (cemilcengiz) [inline README.md:201]: "In this particular class it is needed to transform" -> "This is performed for transforming"

File: README.md
Comment (cemilcengiz) [inline README.md:202]: using BLEU -> while using BLEU

File: README.md
Comment (cemilcengiz) [inline README.md:204]: I don't think it should be the default implementation. The default should be pass-through.

**Review by cemilcengiz — CHANGES_REQUESTED** [review]

File: trapper/metrics/metric.py
Comment (devrimcavusoglu) [reply trapper/metrics/metric.py:1]: I'll look into those implementations first.

File: examples/pos_tagging/experiments/roberta/experiment.jsonnet
Comment (cemilcengiz) [reply examples/pos_tagging/experiments/roberta/experiment.jsonnet:25]: Well, I looked at the code from the main and realized that it was done as a post-process to the metric outputs. I liked your idea, we can implement a method in the base Metric (or MetricHandler) class something like:
`def post_process(score: Dict) -> Dict: return score`
Which returns the score as it is by default.

File: trapper/metrics/__init__.py
Comment (cemilcengiz) [reply trapper/metrics/__init__.py:2]: Yes, unless we specifically want to allow a less qualified access to SquadQuestionAnsweringDataProcessor (or other classes as a general comment). For example, I import both `DataProcessor` and `IndexedInstance` intentionally from `trapper.data.data_processors.data_processor` in the same init file.

File: README.md
Comment (cemilcengiz) [reply README.md]: I am not sure when it is in the beginning of the sentence.

File: trapper/data/metadata_handlers/metadata_handler.py
Comment (devrimcavusoglu) [reply trapper/data/metadata_handlers/metadata_handler.py]: in order to use default metadata handler as is, this is the only reason for not making this method abstract.

File: trapper/metrics/metric_handlers/metric_handler.py
Comment (cemilcengiz) [reply trapper/metrics/metric_handlers/metric_handler.py:15]: We could implement the postprocessing in another class I guess. The two tasks seems unrelated and they do not use / manipulate common instance variables, are they?

File: trapper/data/metadata_handlers/metadata_handler.py
Comment (cemilcengiz) [reply trapper/data/metadata_handlers/metadata_handler.py]: Well, there is another problem with this. Now, by default, we pass a callable that returns none to the dataset.map by this statement `        dataset.map(self.extract_metadata)`. I think, it should return the sample as it is. Moreover, that statement in statement in the _call_ method might be dangerous. You are relying on the self.extract_metadata function to have side effects that does what you think it should do. However, it would be cleaner if we let extract_metadata to always return a type what we expect (if there is any) and do `return dataset.map(self.extract_metadata)` in the _call_ method. However, I am not that familiar to these classes, so I am open for criticism.

File: trapper/metrics/metric_handlers/metric_handler.py
Comment (cemilcengiz) [reply trapper/metrics/metric_handlers/metric_handler.py:76]: I guess argmax() is fine unless you perform some things like decoding (e.g. Viterbi) or beam search etc. However, we do not support these for now.

File: trapper/data/metadata_handlers/metadata_handler.py
Comment (devrimcavusoglu) [reply trapper/data/metadata_handlers/metadata_handler.py]: Here, the intended usage is to have a single `__call__()` method which should not be overridden in any child class (as we discussed before typing.final does support this for python 3.8, but currently no option for that). The intention of the existence of `extract_metadata` is due to to main purpose:
 
1. Provide an interface to `__call__()` method to access dataset instances, and store the required information. Since `__call__()` should not be overridden, we provide an overridable method for child classes.

> However, it would be cleaner if we let extract_metadata to always return a type what we expect (if there is any) and do return dataset.map(self.extract_metadata) in the call method.

2. Second purpose is actually providing us what you mentioned here. If we specify a return type and return an instance we can never guarantee that the instance is manipulated or changed (in any ways). We do not want to have any manipulation on `dataset` passed to `self.extract_metadata()`. Thus, specifying return type as `None` guarantees that `dataset` is not manipulated in any way by the map function, making `self.extract_metadata()` read-only, without a write access. In this method user only extract the metadata required for postprocess of prediction outputs and store them in object variables (attributes), for example like in question answering case.

File: examples/pos_tagging/experiments/roberta/experiment.jsonnet
Comment (devrimcavusoglu) [reply examples/pos_tagging/experiments/roberta/experiment.jsonnet:25]: OK, I'll look into this.

File: trapper/metrics/__init__.py
Comment (devrimcavusoglu) [reply trapper/metrics/__init__.py:2]: OK, I'll remove both `SquadQuestionAnsweringDataProcessor` and `MetricHandlerForQuestionAnswering` from outer init files.

File: trapper/metrics/metric_handlers/metric_handler.py
Comment (devrimcavusoglu) [reply trapper/metrics/metric_handlers/metric_handler.py:15]: They do if required, `postprocessing()` uses attributes stored by `extract_metadata()` e.g in QA case.

File: trapper/metrics/metric_handlers/metric_handler.py
Comment (devrimcavusoglu) [reply trapper/metrics/metric_handlers/metric_handler.py:76]: Yes, of course. I only suggested to keep argmax(). I'll update this way, then.

File: trapper/metrics/metric_handlers/metric_handler.py
Comment (cemilcengiz) [reply trapper/metrics/metric_handlers/metric_handler.py:15]: I don't see how. Btw, QA does not even override the `postprocess`, it uses the base method which returns the score without a change.

Comment (devrimcavusoglu) [reply trapper/metrics/metric_handlers/metric_handler.py:15]: > I don't see how. Btw, QA does not even override the `postprocess()`, it uses the base method which returns the score without a change.

`preprocess` method was named `postprocess` before the changes, it was named `postprocess` because it postprocesses the predictions outputs, but now `postprocess` named `preprocess` as it preprocesses the prediction outputs before metric computation, and current `postprocess` method is for score manipulation (I know it's a bit confusing after the change :sweat_smile: )

File: examples/pos_tagging/src/metric_handler.py
Comment (cemilcengiz) [inline examples/pos_tagging/src/metric_handler.py:45]: This does not replicate the previous behavior, rather it operates like the `return_entity_level_metrics` set to True. Can we add a `return_entity_level_metrics: bool = True` parameter to the constructor just like in the previous implementation?

File: trapper/metrics/metric_handlers/metric_handler.py
Comment (cemilcengiz) [inline trapper/metrics/metric_handlers/metric_handler.py:61]: Line 63 is incorrect, we return None (either explicit or implicitly)

File: trapper/metrics/metric_handlers/metric_handler.py
Comment (cemilcengiz) [inline trapper/metrics/metric_handlers/metric_handler.py:80]: Lets make this method accept `pred: EvalPrediction` and return the same type, to make it more future-compatible with transformers? (Of course for child classes as well).

File: trapper/metrics/metric.py
Comment (cemilcengiz) [inline trapper/metrics/metric.py:19]: Can we split metric.py file into two, namely metric.py and jury.py. Then move the Jury class definiton to jury.py. Actually I made that comment before, but I guess it was missed (or I haven't seen the reply)

File: README.md
Comment (cemilcengiz) [inline README.md:76]: extending the datasets library

File: README.md
Comment (cemilcengiz) [inline README.md:107]: 3 problems.
1) We added the `postprocess` method to this class as well. However, this paragraph is only talking about the `preprocess` method like it is the only (or the main) thing this class performs.
2) `The default behavior of this class is to directly decode` -> This is incorrect since we changed the default behavior. Moreover, problem-1 applies here as well, this is the default behavior of the `preprocess` method, not the class.
3) (Minor one): Instead of `prediction outputs`, let's use somethind like `model predictions`.

File: README.md
Comment (cemilcengiz) [inline README.md:206]: `prediction outputs` -> `model predictions`
`This class is responsible for postprocess operations applied to prediction outputs.` -> This is confusing since the method does that is named `preprocess`. Moreover, as in my previous comment, we added `postprocess` method that does optional postprocessing to the scores. Therefore, this paragraph should also be updated.

**Review by cemilcengiz — CHANGES_REQUESTED** [review]

File: examples/pos_tagging/src/metric_handler.py
Comment (devrimcavusoglu) [reply examples/pos_tagging/src/metric_handler.py:45]: I assumed that it is only implemented for seqeval metric (current implementation on `main`) you want it to be applicable to all metrics ?

File: trapper/metrics/metric_handlers/metric_handler.py
Comment (devrimcavusoglu) [reply trapper/metrics/metric_handlers/metric_handler.py:79]: jury (or in general metric) inputs are named `predictions/references`, changing this to `EvalPrediction` makes it more confusing I think, plus the name implies it like it is only predictions but the object has labels, too. This characteristic of `EvalPrediciton` requires additional knowledge of transformers library. IMHO, I disagree to change the current structure to proposed one for the reasons I mentioned.

File: trapper/metrics/metric_handlers/metric_handler.py
Comment (cemilcengiz) [reply trapper/metrics/metric_handlers/metric_handler.py:15]: I am talking about preprocess vs postprocess, and still not understood how these two methods are expected to operate on the same instance attributes.

File: trapper/metrics/metric_handlers/metric_handler.py
Comment (cemilcengiz) [reply trapper/metrics/metric_handlers/metric_handler.py:79]: I think, we would just be hiding by leaving them as tuple of predictions, references instead of the `EvalPrediction` type. The reason is that, they are actually the `EvalPrediction` in disguise. If for a reason, the transformers library changed the object returned by the model.forward(), we would have to change these metrics as well. 
In fact, requiring the knowledge of `EvalPrediciton` type is fine, that is what we use already. The models in trapper returns that type (and only that type), so it is only natural to have the metrics use that type.

edit: BTW, we are implementing the features for the trapper, which is built on top of transformers and supposes complete integration. If we were talking about the Jury metric interfaces in the jury library itself, I would not oppose.

File: trapper/metrics/metric_handlers/metric_handler.py
Comment (cemilcengiz) [inline trapper/metrics/metric_handlers/metric_handler.py:96]: Again, I think we use used return `EvalPrediction`.

File: trapper/metrics/metric_handlers/question_answering_handler.py
Comment (cemilcengiz) [inline trapper/metrics/metric_handlers/question_answering_handler.py:13]: `converting the` -> `the conversion of` maybe?
begin -> `begin` or `the beginning`,
end -> `end` or `the end`

File: README.md
Comment (cemilcengiz) [inline README.md:75]: Can we replace this:
`We implement the trapper components by trying to use the available components of the transformers library as much as we can. `
with
`While implementing the components of trapper, we try to reuse the classes from the transformers library as much as we can.`

File: README.md
Comment (cemilcengiz) [inline README.md:215]: This sentence should be fixed, at least starting from `so before directly `

File: examples/pos_tagging/src/metric_handler.py
Comment (cemilcengiz) [reply examples/pos_tagging/src/metric_handler.py:45]: Let's move this class to trapper with the change I recommended as well as renaming it to SeqevalMetricHandler.

File: examples/pos_tagging/src/metric_handler.py
Comment (cemilcengiz) [inline examples/pos_tagging/src/metric_handler.py:19]: `SeqEvalMetric`  should be changed.

File: trapper/metrics/metric_handlers/metric_handler.py
Comment (cemilcengiz) [inline trapper/metrics/metric_handlers/metric_handler.py:91]: why not just use `model predictions`. I still find `prediction outputs` confusing, it is like we are talking about a predictive model.

File: README.md
Comment (cemilcengiz) [inline README.md:213]: `prediction outputs` requires change here as well.

**Review by cemilcengiz — CHANGES_REQUESTED** [review]

File: README.md
Comment (devrimcavusoglu) [reply README.md:215]: I rewrote that part.

**Review by cemilcengiz — COMMENTED** [review]
LGTM. After I update the README, we can merge this.

**Pull request merged**
